Reflect::hash collides with Hash::hash

With the merging of @cart's #926, the `hash` method of `bevy::prelude::Reflect` now collides with the `hash` method from `std::hash::Hash`.  You can workaround by disambiguating, so this isn't a showstopper.  I'm reporting it in the hopes that there's an easy tweak to avoid having to do the workaround. If not, no big deal - and I'm fine closing this.

With the following imports:

```rust
use bevy::prelude::*;
use std::hash::Hash;
```

When you attempt to call the `hash` method on a primitive value like an integer (which is something you might want to do within an implementation of the Hash trait for a struct, such as on line 59 here -- `PlayerID` is a newtype over `usize`):

https://github.com/CleanCut/punchball/blob/9f8ddb2de2b8a91629307ee81b3f64aaa1eed1de/src/player/collision.rs#L52-L62

You get this error:

```
error[E0034]: multiple applicable items in scope
  --> src/player/collision.rs:58:23
   |
58 |             player_id.hash(state);
   |                       ^^^^ multiple `hash` found
   |
   = note: candidate #1 is defined in an impl of the trait `Hash` for the type `usize`
   = note: candidate #2 is defined in an impl of the trait `bevy::prelude::Reflect` for the type `usize`
help: disambiguate the associated function for candidate #1
   |
58 |             Hash::hash(&player_id, state);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: disambiguate the associated function for candidate #2
   |
58 |             bevy::prelude::Reflect::hash(&player_id, state);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0034`.
```

The workaround is to change `player_id.hash(state);` to `Hash::hash(&player_id, state);` to disambiguate.

https://github.com/CleanCut/punchball/blob/a6abd16d239fb7c6ea2656990c6da16056d4ffce/src/player/collision.rs#L52-L62

--- a/crates/bevy_reflect/README.md
+++ b/crates/bevy_reflect/README.md
@@ -102,7 +102,7 @@ let reflect_deserializer = ReflectDeserializer::new(&registry);
 let value = reflect_deserializer.deserialize(&mut deserializer).unwrap();
 let dynamic_struct = value.take::<DynamicStruct>().unwrap();
 
-assert!(foo.partial_eq(&dynamic_struct).unwrap());
+assert!(foo.reflect_partial_eq(&dynamic_struct).unwrap());
 ```
 
 ### Trait "reflection"

--- a/crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
+++ b/crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
@@ -311,7 +311,7 @@ fn impl_struct(
                 #serialize_fn
             }
 
-            fn hash(&self) -> Option<u64> {
+            fn reflect_hash(&self) -> Option<u64> {
                 #hash_fn
             }
 
@@ -430,7 +430,7 @@ fn impl_tuple_struct(
                 #serialize_fn
             }
 
-            fn hash(&self) -> Option<u64> {
+            fn reflect_hash(&self) -> Option<u64> {
                 #hash_fn
             }
 
@@ -501,7 +501,7 @@ fn impl_value(
                 #bevy_reflect_path::ReflectMut::Value(self)
             }
 
-            fn hash(&self) -> Option<u64> {
+            fn reflect_hash(&self) -> Option<u64> {
                 #hash_fn
             }
 
@@ -585,7 +585,7 @@ pub fn impl_reflect_value(input: TokenStream) -> TokenStream {
 
 #[derive(Default)]
 struct ReflectAttrs {
-    hash: TraitImpl,
+    reflect_hash: TraitImpl,
     partial_eq: TraitImpl,
     serialize: TraitImpl,
     data: Vec<Ident>,
@@ -603,7 +603,7 @@ impl ReflectAttrs {
                             let ident = segment.ident.to_string();
                             match ident.as_str() {
                                 "PartialEq" => attrs.partial_eq = TraitImpl::Implemented,
-                                "Hash" => attrs.hash = TraitImpl::Implemented,
+                                "Hash" => attrs.reflect_hash = TraitImpl::Implemented,
                                 "Serialize" => attrs.serialize = TraitImpl::Implemented,
                                 _ => attrs.data.push(Ident::new(
                                     &format!("Reflect{}", segment.ident),
@@ -630,7 +630,7 @@ impl ReflectAttrs {
                                                         TraitImpl::Custom(segment.ident.clone())
                                                 }
                                                 "Hash" => {
-                                                    attrs.hash =
+                                                    attrs.reflect_hash =
                                                         TraitImpl::Custom(segment.ident.clone())
                                                 }
                                                 "Serialize" => {
@@ -657,7 +657,7 @@ impl ReflectAttrs {
     }
 
     fn get_hash_impl(&self, path: &Path) -> proc_macro2::TokenStream {
-        match &self.hash {
+        match &self.reflect_hash {
             TraitImpl::Implemented => quote! {
                 use std::hash::{Hash, Hasher};
                 let mut hasher = #path::ReflectHasher::default();

--- a/crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
+++ b/crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
@@ -196,7 +196,7 @@ fn impl_struct(
 
     let hash_fn = reflect_attrs.get_hash_impl(&bevy_reflect_path);
     let serialize_fn = reflect_attrs.get_serialize_impl(&bevy_reflect_path);
-    let partial_eq_fn = match reflect_attrs.partial_eq {
+    let partial_eq_fn = match reflect_attrs.reflect_partial_eq {
         TraitImpl::NotImplemented => quote! {
             use #bevy_reflect_path::Struct;
             #bevy_reflect_path::struct_partial_eq(self, value)
@@ -315,7 +315,7 @@ fn impl_struct(
                 #hash_fn
             }
 
-            fn partial_eq(&self, value: &dyn #bevy_reflect_path::Reflect) -> Option<bool> {
+            fn reflect_partial_eq(&self, value: &dyn #bevy_reflect_path::Reflect) -> Option<bool> {
                 #partial_eq_fn
             }
         }
@@ -339,7 +339,7 @@ fn impl_tuple_struct(
 
     let hash_fn = reflect_attrs.get_hash_impl(&bevy_reflect_path);
     let serialize_fn = reflect_attrs.get_serialize_impl(&bevy_reflect_path);
-    let partial_eq_fn = match reflect_attrs.partial_eq {
+    let partial_eq_fn = match reflect_attrs.reflect_partial_eq {
         TraitImpl::NotImplemented => quote! {
             use #bevy_reflect_path::TupleStruct;
             #bevy_reflect_path::tuple_struct_partial_eq(self, value)
@@ -434,7 +434,7 @@ fn impl_tuple_struct(
                 #hash_fn
             }
 
-            fn partial_eq(&self, value: &dyn #bevy_reflect_path::Reflect) -> Option<bool> {
+            fn reflect_partial_eq(&self, value: &dyn #bevy_reflect_path::Reflect) -> Option<bool> {
                 #partial_eq_fn
             }
         }
@@ -505,7 +505,7 @@ fn impl_value(
                 #hash_fn
             }
 
-            fn partial_eq(&self, value: &dyn #bevy_reflect_path::Reflect) -> Option<bool> {
+            fn reflect_partial_eq(&self, value: &dyn #bevy_reflect_path::Reflect) -> Option<bool> {
                 #partial_eq_fn
             }
 
@@ -586,7 +586,7 @@ pub fn impl_reflect_value(input: TokenStream) -> TokenStream {
 #[derive(Default)]
 struct ReflectAttrs {
     reflect_hash: TraitImpl,
-    partial_eq: TraitImpl,
+    reflect_partial_eq: TraitImpl,
     serialize: TraitImpl,
     data: Vec<Ident>,
 }
@@ -602,7 +602,7 @@ impl ReflectAttrs {
                         if let Some(segment) = path.segments.iter().next() {
                             let ident = segment.ident.to_string();
                             match ident.as_str() {
-                                "PartialEq" => attrs.partial_eq = TraitImpl::Implemented,
+                                "PartialEq" => attrs.reflect_partial_eq = TraitImpl::Implemented,
                                 "Hash" => attrs.reflect_hash = TraitImpl::Implemented,
                                 "Serialize" => attrs.serialize = TraitImpl::Implemented,
                                 _ => attrs.data.push(Ident::new(
@@ -626,7 +626,7 @@ impl ReflectAttrs {
                                         if let Some(segment) = path.segments.iter().next() {
                                             match ident.as_str() {
                                                 "PartialEq" => {
-                                                    attrs.partial_eq =
+                                                    attrs.reflect_partial_eq =
                                                         TraitImpl::Custom(segment.ident.clone())
                                                 }
                                                 "Hash" => {
@@ -675,7 +675,7 @@ impl ReflectAttrs {
     }
 
     fn get_partial_eq_impl(&self) -> proc_macro2::TokenStream {
-        match &self.partial_eq {
+        match &self.reflect_partial_eq {
             TraitImpl::Implemented => quote! {
                 let value = value.any();
                 if let Some(value) = value.downcast_ref::<Self>() {

--- a/crates/bevy_reflect/src/impls/smallvec.rs
+++ b/crates/bevy_reflect/src/impls/smallvec.rs
@@ -82,7 +82,7 @@ where
         Box::new(self.clone_dynamic())
     }
 
-    fn hash(&self) -> Option<u64> {
+    fn reflect_hash(&self) -> Option<u64> {
         None
     }
 

--- a/crates/bevy_reflect/src/impls/smallvec.rs
+++ b/crates/bevy_reflect/src/impls/smallvec.rs
@@ -86,7 +86,7 @@ where
         None
     }
 
-    fn partial_eq(&self, value: &dyn Reflect) -> Option<bool> {
+    fn reflect_partial_eq(&self, value: &dyn Reflect) -> Option<bool> {
         crate::list_partial_eq(self, value)
     }
 

--- a/crates/bevy_reflect/src/impls/std.rs
+++ b/crates/bevy_reflect/src/impls/std.rs
@@ -97,7 +97,7 @@ impl<T: Reflect> Reflect for Vec<T> {
         None
     }
 
-    fn partial_eq(&self, value: &dyn Reflect) -> Option<bool> {
+    fn reflect_partial_eq(&self, value: &dyn Reflect) -> Option<bool> {
         crate::list_partial_eq(self, value)
     }
 
@@ -191,7 +191,7 @@ impl<K: Reflect + Clone + Eq + Hash, V: Reflect + Clone> Reflect for HashMap<K, 
         None
     }
 
-    fn partial_eq(&self, value: &dyn Reflect) -> Option<bool> {
+    fn reflect_partial_eq(&self, value: &dyn Reflect) -> Option<bool> {
         map_partial_eq(self, value)
     }
 

--- a/crates/bevy_reflect/src/impls/std.rs
+++ b/crates/bevy_reflect/src/impls/std.rs
@@ -93,7 +93,7 @@ impl<T: Reflect> Reflect for Vec<T> {
         Box::new(self.clone_dynamic())
     }
 
-    fn hash(&self) -> Option<u64> {
+    fn reflect_hash(&self) -> Option<u64> {
         None
     }
 
@@ -187,7 +187,7 @@ impl<K: Reflect + Clone + Eq + Hash, V: Reflect + Clone> Reflect for HashMap<K, 
         Box::new(self.clone_dynamic())
     }
 
-    fn hash(&self) -> Option<u64> {
+    fn reflect_hash(&self) -> Option<u64> {
         None
     }
 

--- a/crates/bevy_reflect/src/lib.rs
+++ b/crates/bevy_reflect/src/lib.rs
@@ -306,7 +306,7 @@ mod tests {
         let value = reflect_deserializer.deserialize(&mut deserializer).unwrap();
         let dynamic_struct = value.take::<DynamicStruct>().unwrap();
 
-        assert!(foo.partial_eq(&dynamic_struct).unwrap());
+        assert!(foo.reflect_partial_eq(&dynamic_struct).unwrap());
     }
 
     #[test]

--- a/crates/bevy_reflect/src/list.rs
+++ b/crates/bevy_reflect/src/list.rs
@@ -111,7 +111,7 @@ impl Reflect for DynamicList {
     }
 
     #[inline]
-    fn hash(&self) -> Option<u64> {
+    fn reflect_hash(&self) -> Option<u64> {
         None
     }
 

--- a/crates/bevy_reflect/src/list.rs
+++ b/crates/bevy_reflect/src/list.rs
@@ -115,7 +115,7 @@ impl Reflect for DynamicList {
         None
     }
 
-    fn partial_eq(&self, value: &dyn Reflect) -> Option<bool> {
+    fn reflect_partial_eq(&self, value: &dyn Reflect) -> Option<bool> {
         list_partial_eq(self, value)
     }
 
@@ -169,7 +169,7 @@ pub fn list_partial_eq<L: List>(a: &L, b: &dyn Reflect) -> Option<bool> {
     }
 
     for (a_value, b_value) in a.iter().zip(list.iter()) {
-        if let Some(false) | None = a_value.partial_eq(b_value) {
+        if let Some(false) | None = a_value.reflect_partial_eq(b_value) {
             return Some(false);
         }
     }

--- a/crates/bevy_reflect/src/map.rs
+++ b/crates/bevy_reflect/src/map.rs
@@ -134,7 +134,7 @@ impl Reflect for DynamicMap {
         None
     }
 
-    fn partial_eq(&self, value: &dyn Reflect) -> Option<bool> {
+    fn reflect_partial_eq(&self, value: &dyn Reflect) -> Option<bool> {
         map_partial_eq(self, value)
     }
 
@@ -172,7 +172,7 @@ pub fn map_partial_eq<M: Map>(a: &M, b: &dyn Reflect) -> Option<bool> {
 
     for (key, value) in a.iter() {
         if let Some(map_value) = map.get(key) {
-            if let Some(false) | None = value.partial_eq(map_value) {
+            if let Some(false) | None = value.reflect_partial_eq(map_value) {
                 return Some(false);
             }
         } else {

--- a/crates/bevy_reflect/src/map.rs
+++ b/crates/bevy_reflect/src/map.rs
@@ -33,7 +33,7 @@ impl DynamicMap {
     }
 
     pub fn insert_boxed(&mut self, key: Box<dyn Reflect>, value: Box<dyn Reflect>) {
-        match self.indices.entry(key.hash().expect(HASH_ERROR)) {
+        match self.indices.entry(key.reflect_hash().expect(HASH_ERROR)) {
             Entry::Occupied(entry) => {
                 self.values[*entry.get()] = (key, value);
             }
@@ -48,13 +48,13 @@ impl DynamicMap {
 impl Map for DynamicMap {
     fn get(&self, key: &dyn Reflect) -> Option<&dyn Reflect> {
         self.indices
-            .get(&key.hash().expect(HASH_ERROR))
+            .get(&key.reflect_hash().expect(HASH_ERROR))
             .map(|index| &*self.values.get(*index).unwrap().1)
     }
 
     fn get_mut(&mut self, key: &dyn Reflect) -> Option<&mut dyn Reflect> {
         self.indices
-            .get(&key.hash().expect(HASH_ERROR))
+            .get(&key.reflect_hash().expect(HASH_ERROR))
             .cloned()
             .map(move |index| &mut *self.values.get_mut(index).unwrap().1)
     }
@@ -130,7 +130,7 @@ impl Reflect for DynamicMap {
         Box::new(self.clone_dynamic())
     }
 
-    fn hash(&self) -> Option<u64> {
+    fn reflect_hash(&self) -> Option<u64> {
         None
     }
 

--- a/crates/bevy_reflect/src/reflect.rs
+++ b/crates/bevy_reflect/src/reflect.rs
@@ -32,7 +32,7 @@ pub trait Reflect: Any + Send + Sync {
     /// Returns a hash of the value (which includes the type) if hashing is supported. Otherwise `None` will be returned.
     fn reflect_hash(&self) -> Option<u64>;
     /// Returns a "partial equal" comparison result if comparison is supported. Otherwise `None` will be returned.
-    fn partial_eq(&self, _value: &dyn Reflect) -> Option<bool>;
+    fn reflect_partial_eq(&self, _value: &dyn Reflect) -> Option<bool>;
     /// Returns a serializable value, if serialization is supported. Otherwise `None` will be returned.
     fn serializable(&self) -> Option<Serializable>;
 }

--- a/crates/bevy_reflect/src/reflect.rs
+++ b/crates/bevy_reflect/src/reflect.rs
@@ -30,7 +30,7 @@ pub trait Reflect: Any + Send + Sync {
     fn reflect_mut(&mut self) -> ReflectMut;
     fn clone_value(&self) -> Box<dyn Reflect>;
     /// Returns a hash of the value (which includes the type) if hashing is supported. Otherwise `None` will be returned.
-    fn hash(&self) -> Option<u64>;
+    fn reflect_hash(&self) -> Option<u64>;
     /// Returns a "partial equal" comparison result if comparison is supported. Otherwise `None` will be returned.
     fn partial_eq(&self, _value: &dyn Reflect) -> Option<bool>;
     /// Returns a serializable value, if serialization is supported. Otherwise `None` will be returned.

--- a/crates/bevy_reflect/src/struct_trait.rs
+++ b/crates/bevy_reflect/src/struct_trait.rs
@@ -216,7 +216,7 @@ impl Reflect for DynamicStruct {
         Ok(())
     }
 
-    fn hash(&self) -> Option<u64> {
+    fn reflect_hash(&self) -> Option<u64> {
         None
     }
 

--- a/crates/bevy_reflect/src/struct_trait.rs
+++ b/crates/bevy_reflect/src/struct_trait.rs
@@ -220,7 +220,7 @@ impl Reflect for DynamicStruct {
         None
     }
 
-    fn partial_eq(&self, value: &dyn Reflect) -> Option<bool> {
+    fn reflect_partial_eq(&self, value: &dyn Reflect) -> Option<bool> {
         struct_partial_eq(self, value)
     }
 
@@ -244,7 +244,7 @@ pub fn struct_partial_eq<S: Struct>(a: &S, b: &dyn Reflect) -> Option<bool> {
     for (i, value) in struct_value.iter_fields().enumerate() {
         let name = struct_value.name_at(i).unwrap();
         if let Some(field_value) = a.field(name) {
-            if let Some(false) | None = field_value.partial_eq(value) {
+            if let Some(false) | None = field_value.reflect_partial_eq(value) {
                 return Some(false);
             }
         } else {

--- a/crates/bevy_reflect/src/tuple_struct.rs
+++ b/crates/bevy_reflect/src/tuple_struct.rs
@@ -171,7 +171,7 @@ impl Reflect for DynamicTupleStruct {
         Ok(())
     }
 
-    fn hash(&self) -> Option<u64> {
+    fn reflect_hash(&self) -> Option<u64> {
         None
     }
 

--- a/crates/bevy_reflect/src/tuple_struct.rs
+++ b/crates/bevy_reflect/src/tuple_struct.rs
@@ -175,7 +175,7 @@ impl Reflect for DynamicTupleStruct {
         None
     }
 
-    fn partial_eq(&self, value: &dyn Reflect) -> Option<bool> {
+    fn reflect_partial_eq(&self, value: &dyn Reflect) -> Option<bool> {
         tuple_struct_partial_eq(self, value)
     }
 
@@ -198,7 +198,7 @@ pub fn tuple_struct_partial_eq<S: TupleStruct>(a: &S, b: &dyn Reflect) -> Option
 
     for (i, value) in tuple_struct.iter_fields().enumerate() {
         if let Some(field_value) = a.field(i) {
-            if let Some(false) | None = field_value.partial_eq(value) {
+            if let Some(false) | None = field_value.reflect_partial_eq(value) {
                 return Some(false);
             }
         } else {

--- a/examples/reflection/reflection.rs
+++ b/examples/reflection/reflection.rs
@@ -84,9 +84,9 @@ fn setup(type_registry: Res<TypeRegistry>) {
     // as themselves.
     let _deserialized_struct = reflect_value.downcast_ref::<DynamicStruct>();
 
-    // Reflect has its own `partial_eq` implementation. This behaves like normal `partial_eq`, but it treats "dynamic" and
+    // Reflect has its own `partial_eq` implementation, named `reflect_partial_eq`. This behaves like normal `partial_eq`, but it treats "dynamic" and
     // "non-dynamic" types the same. The `Foo` struct and deserialized `DynamicStruct` are considered equal for this reason:
-    assert!(reflect_value.partial_eq(&value).unwrap());
+    assert!(reflect_value.reflect_partial_eq(&value).unwrap());
 
     // By "patching" `Foo` with the deserialized DynamicStruct, we can "Deserialize" Foo.
     // This means we can serialize and deserialize with a single `Reflect` derive!

